packages/main/src/Toolbar.ts

```diff
@@ -0,0 +1,562 @@
+
+	/**
+	 * Items, which will be displayed in the toolbar.
+	 * @type {Object}
+	 * @private
+	 */
+	@property({ type: Object, multiple: true })
+	itemsToBar!: Array<ToolbarItem>;
+	/**
+	 * Items, that will be displayed inside overflow Popover.
+	 * @type {Object}
+	 * @private
+	 */
+	@property({ type: Object, multiple: true })
+	itemsToOverflow!: Array<ToolbarItem>;
+
+	/**
+	 * Cached the sum of all of items width.
+	 * @type {Object}
+	 * @private
+	 */
+	 @property({ type: Integer })
+	 ITEMS_WIDTH: number;
+
+	/**
+	 * Indicates the items have been measured and the layout can be calculated.
+	 * @type {boolean}
+	 * @private
+	 */
+	@property({ type: Boolean })
+	itemsWidthMeasured!: boolean;
+	/**
+	 * Indicates the end of the resizing iteration.
+	 * @type {boolean}
+	 * @private
+	 */
+	@property({ type: Boolean })
+	resizing!: boolean;
+
+	/**
+	* Calculated width of the whole toolbar.
+	* @private
+	*/
+	@property({ type: Integer })
+	width!: number;
+
+	/**
+	* @private
+	* Calculated width of all the Toolbar items.
+	*/
+	@property({ type: Integer })
+	contentWidth!: number;
+
+	/**
+	 * Notifies the toolbar if it should show the items in a reverse way if Toolbar Popover needs to be placed on "Top" position.
+	 * @private
+	 * @type {Boolean}
+	 */
+	@property({ type: Boolean })
+	reverseOverflow!: boolean;
+
+	/**
+ 	* Slotted Toolbar items
+	* @type {sap.ui.webc.main.IToolbarItem[]}
+	* @name sap.ui.webc.main.Toolbar.prototype.items
+	* @slot items
+	* @public
+	*/
+	@slot({ "default": true, type: HTMLElement, invalidateOnChildChange: true })
+	items!: Array<ToolbarItem>
+
+	_onResize!: ResizeObserverCallback;
+
+	ITEMS_WIDTH_MAP: Map<string, number>;
+
+	constructor() {
+		super();
+
+		// indicates if the bar is being reized at the moment
+		this.resizing = false;
+
+		this.ITEMS_WIDTH_MAP = new Map();
+
+		this.ITEMS_WIDTH = 0;
+
+		// resize handler
+		this._onResize = this.onResize.bind(this);
+	}
+
+	get OVERFLOW_BTN_SIZE(): number {
+		const toolbarOverflowButton = this.shadowRoot!.querySelector(".ui5-tb-overflow-btn")!;
+		return toolbarOverflowButton ? toolbarOverflowButton.getBoundingClientRect().width : 0;
+	}
+
+	get PADDING(): number {
+		const toolbarComputedStyle = getComputedStyle(this.getDomRef()!);
+		return this.calculateCSSREMValue(toolbarComputedStyle, "--_ui5-toolbar-padding-left")
+		+ this.calculateCSSREMValue(toolbarComputedStyle, "--_ui5-toolbar-padding-right");
+	}
+
+	/**
+	 * Lifecycle methods
+	 */
+	onEnterDOM() {
+		ResizeHandler.register(this, this._onResize);
+	}
+
+	onExitDOM() {
+		ResizeHandler.deregister(this, this._onResize);
+	}
+
+	async onAfterRendering() {
+		if (this.resizing) {
+			this.resizing = false;
+			return;
+		}
+
+		await renderFinished();
+		this.storeItemsWidth();
+		this.processOverflowLayout();
+	}
+
+	calculateCSSREMValue(styleSet: CSSStyleDeclaration, propertyName: string): number {
+		return Number(styleSet.getPropertyValue(propertyName).replace("rem", "")) * 16;
```

Comment:
we can't use 16 as hardcoded value here because 1 rem size depends on the body font-size which could be changed